go/store/datas/pull/pull.go

```diff
@@ -0,0 +1,236 @@
+		updateProgress(0, uint64(absentCount), 0)
+
+		// For gathering up the hashes in the next level of the tree
+		nextLevel := hash.HashSet{}
+		uniqueOrdered := hash.HashSlice{}
+
+		// Process all absent chunks in this level of the tree in quanta of at most |batchSize|
+		for start, end := 0, batchSize; start < absentCount; start, end = end, end+batchSize {
+			if end > absentCount {
+				end = absentCount
+			}
+			batch := absent[start:end]
+
+			neededChunks, err := getChunks(ctx, srcCS, batch, sampleSize, sampleCount, updateProgress)
+
+			if err != nil {
+				return err
+			}
+
+			uniqueOrdered, err = putChunks(ctx, walkRefs, sinkCS, batch, neededChunks, nextLevel, uniqueOrdered)
+
+			if err != nil {
+				return err
+			}
+		}
+
+		absent, err = nextLevelMissingChunks(ctx, sinkCS, nextLevel, absent, uniqueOrdered)
+
+		if err != nil {
+			return err
+		}
+	}
+
+	err = persistChunks(ctx, sinkCS)
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func persistChunks(ctx context.Context, cs chunks.ChunkStore) error {
+	// todo: there is no call to rebase on an unsuccessful Commit()
+	// will  this loop forever?
+	var success bool
+	for !success {
+		r, err := cs.Root(ctx)
+
+		if err != nil {
+			return err
+		}
+
+		success, err = cs.Commit(ctx, r, r)
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// PullWithoutBatching effectively removes the batching of chunk retrieval done on each level of the tree.  This means
+// all chunks from one level of the tree will be retrieved from the underlying chunk store in one call, which pushes the
+// optimization problem down to the chunk store which can make smarter decisions.
+func PullWithoutBatching(ctx context.Context, srcCS, sinkCS chunks.ChunkStore, walkRefs WalkRefs, sourceHash hash.Hash, progressCh chan PullProgress) error {
+	// by increasing the batch size to MaxInt32 we effectively remove batching here.
+	return pull(ctx, srcCS, sinkCS, walkRefs, sourceHash, progressCh, math.MaxInt32)
+}
+
+// concurrently pull all chunks from this batch that the sink is missing out of the source
+func getChunks(ctx context.Context, srcCS chunks.ChunkStore, batch hash.HashSlice, sampleSize uint64, sampleCount uint64, updateProgress func(moreDone uint64, moreKnown uint64, moreApproxBytesWritten uint64)) (map[hash.Hash]*chunks.Chunk, error) {
+	mu := &sync.Mutex{}
+	neededChunks := map[hash.Hash]*chunks.Chunk{}
+	err := srcCS.GetMany(ctx, batch.HashSet(), func(ctx context.Context, c *chunks.Chunk) {
+		mu.Lock()
+		defer mu.Unlock()
+		neededChunks[c.Hash()] = c
+
+		// Randomly sample amount of data written
+		if rand.Float64() < bytesWrittenSampleRate {
+			sampleSize += uint64(len(snappy.Encode(nil, c.Data())))
+			sampleCount++
+		}
+		updateProgress(1, 0, sampleSize/uint64(math.Max(1, float64(sampleCount))))
+	})
+	if err != nil {
+		return nil, err
+	}
+	return neededChunks, nil
+}
+
+type WalkRefs func(chunks.Chunk, func(hash.Hash, uint64) error) error
```

Comment:
Do we ever need the `height uint64` param? It's not currently available in the new format

Reply:
AFAICT, the existing stuff could be adapted to not need it. It's currently used for reporting how many levels of the tree we have left to process and for keeping track of leaves so that we don't WalkRefs them.

In the new format, I think both of those use cases can be satisfied basedon the self-reported height of the chunk itself.

In the past, `jankyfetch` used the height to aggressively attack the highest refs before going after the lower portions of the DAG. A similar effect could probably be had by using the height of the node the reference is in as the heuristic.